### ISSUE TITLE
Getting top-k similar images from the network embeddings for each input image ?

### ISSUE THREAD

**Issue by ArtiShelar** [cvjena/semantic-embeddings#5]
Hello there!
We are working on a small project, where our problem statement is to extract the top-k similar images from the network trained on the ImageNet dataset, given 1 input image. We found your Github repo and paper during our research, and thought that hierarchical image embeddings might give better results for fetching similar images, as compared to a network that is pre-trained only on ImageNet (without hierarchical learning). 

We tried working with the pre-trained models provided by you, but due to compatibility issues with TF 2.x, we are unable to use them as is. We are currently working on upgrading a part of this repo, to suit our needs.

We would like to train ResNet50 on ImageNet, to have approximately the same accuracy as your pre-trained models. So, we had a couple of questions.

1. As per the readme, we first need to execute compute_class_embeddings.py and then learn_image_embeddings.py. After doing this, and training for 80 epochs, the accuracy is still not up to our needs. Is the execution of learn_labelembeddings.py required for better accuracy ?  Is there something that we are missing ?
2. Once the network is trained well, our plan is to generate embeddings for each image from training set, and using it to create an index for efficient similarity search for top-k images, given any 1 input image. With this technique we are currently not receiving images from the same class as the input image, in the top-10 results. What technique was used to receive the results that are shown in the first image in the readme file ?

**Comment by Callidior** [comment]
Hey,
I am glad to hear about your interest in our work.

I think a lot of issues can arise alone from upgrading to TF 2.0. If possible, I would recommend installing TF 1.8 and Keras 2.2 in a separate environment and first trying to reproduce the results using that. `learn_labelembeddings.py` implements another method that we used as baseline, so running `learn_image_embeddings.py` is enough. Could you specify the entire command including the arguments you are using? To be on the safe side, it would also be helpful to know how you ran `compute_class_embeddings.py´.

For retrieval, we compute the embeddings using the trained networks for all dataset images and the query image. Since the embeddings are L2-normalized, the cosine similarity can be computed by a simple and efficient dot-product operation. Dataset images are then ranked by decreasing dot-product similarity.
